### PR TITLE
Recall last live action.

### PR DESCRIPTION
**Pull request by KvelaGorrrrnio**
Adds action for recalling the last taken live part.

### CODE REVIEW

File: src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts [hunk 1]
```diff
@@ -17,15 +23,23 @@ import {
...
   ) {}
 
   public createRemoteActions(blueprintConfiguration: Tv2BlueprintConfiguration): Tv2RemoteAction[] {
+    return [
+      ...this.createInsertInsertRemoteAsNextActions(blueprintConfiguration),
+      this.createRecallLastRemoteAsNextAction(),
+    ]
+  }
+
+  private createInsertInsertRemoteAsNextActions(blueprintConfiguration: Tv2BlueprintConfiguration): Tv2RemoteAction[] {
```
Comment (LindvedKrvang) [inline src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts:42]: Insertception :D

File: src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts [hunk 2]
```diff
@@ -116,4 +131,108 @@ export class Tv2RemoteActionFactory {
...
+      },
+      data: {
+        partInterface: {} as PartInterface, // Is determined when called.
+        pieceInterfaces: [] as PieceInterface[], // Is determined when called.
+      }
+    }
+  }
+
+  public isRemoteAction(action: Action): action is Tv2RemoteAction {
+    const tv2Action: Tv2Action = action as Tv2Action
```
Comment (LindvedKrvang) [inline src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts:154]: You can just have the parameter be `Tv2Action`. Then you don't need to cast :)

File: src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts [hunk 3]
```diff
@@ -116,4 +131,108 @@ export class Tv2RemoteActionFactory {
...
+    return [
+      {
+        type: MutateActionType.HISTORIC_PART,
+        updateActionWithPartData: this.updateInsertLastVideoClipToInputAction.bind(this),
+        partPredicate: (part: Part) => this.doesPartContainARemotePiece(part),
+      }
+    ]
+  }
+
+  private updateInsertLastVideoClipToInputAction(action: Action, historicPart: Part): Tv2RecallLastRemoteAsNextAction {
```
Comment (LindvedKrvang) [inline src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts:178]: I think you forgot to update the name :)

File: src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts [hunk 4]
```diff
@@ -116,4 +131,108 @@ export class Tv2RemoteActionFactory {
...
+  }
+
+  private doesPartContainARemotePiece(part: Part): boolean {
+    return part.getPieces().some(this.isRemotePiece.bind(this))
+  }
+
+  private isRemotePiece(piece: Piece): boolean {
+    const metadata: Tv2PieceMetadata | undefined = piece.metadata as Tv2PieceMetadata | undefined
+    return metadata?.type === Tv2PieceType.REMOTE
+  }
```
Comment (LindvedKrvang) [inline src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts:237]: Do you really need to bind `this`? You are not using `this` in `isRemotePiece`.

Comment (LindvedKrvang) [inline src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts:237]: We are only allowed to recall the last _planned_ Live.
I don't see you checking for that anywhere?

File: src/blueprints/tv2/value-objects/tv2-action.ts [hunk 5]
```diff
@@ -60,6 +61,13 @@ export interface Tv2RemoteAction extends Tv2PartAction {
   }
 }
 
+export interface Tv2RecallLastRemoteAsNextAction extends Tv2RemoteAction {
+  type: PartActionType.INSERT_PART_AS_NEXT,
+  metadata: Tv2RemoteAction['metadata'] & {
+    actionSubtype: Tv2ActionSubtype.RECALL_LAST_REMOTE,
+  }
```
Comment (LindvedKrvang) [inline src/blueprints/tv2/value-objects/tv2-action.ts:68]: Couldn't this just be:
```
metadata: {
  actionSubtype: Tv2ActionSubtype:RECALL_LAST_REMOTE
}
```

You set the value of `remoteNumber` to `-1` with the comment `// Is not used for this action`. 
If the attributes of `Tv2RemoteAction['metada']` is not used for this Action, then why do the type gymnastics to include them?

File: src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts (hunk 1 above)
Comment (KvelaGorrrrnio) [reply src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts:42]: Corrected :D

File: src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts (hunk 2 above)
Comment (KvelaGorrrrnio) [reply src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts:154]: Thanks – updated.

File: src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts (hunk 3 above)
Comment (KvelaGorrrrnio) [reply src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts:178]: I don't know why I wrote this :D But it has been updated.

File: src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts (hunk 4 above)
Comment (KvelaGorrrrnio) [reply src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts:237]: Lambda is used instead.

Comment (KvelaGorrrrnio) [reply src/blueprints/tv2/action-factories/tv2-remote-action-factory.ts:237]: Check has been added.

File: src/blueprints/tv2/value-objects/tv2-action.ts (hunk 5 above)
Comment (KvelaGorrrrnio) [reply src/blueprints/tv2/value-objects/tv2-action.ts]: `contentType` is in that case missing. But you are right. I have made it extend from Tv2Part instead.

**Review by LindvedKrvang — APPROVED** [review]

**Review by RasmusAlbrektsen — APPROVED** [review]

**Pull request merged**